doc/adr/0001-jetstream-json-api-design.md

```diff
@@ -0,0 +1,134 @@
+# 1. JetStream JSON API Design
+
+Date: 2020-04-30
+
+## Status
+
+Proposed
+
+## Context
+
+At present, the API encoding consists of mixed text and JSON, we should improve consistency and error handling.
+
+## Decision
+
+We anticipate many types of consumer for JetStream including very small devices via MQTT and those written in 
+languages without easy to use JSON support like C.
+
+We will make it easy for light-weight clients by using `+OK` and `-ERR` style prefixes for client interactions,
+but the server management API will be JSON all the time. The rationale being that such small devices will most
+likely have their Consumers and Streams provisioned as part of the platform and often might not care for the
+additional context.
+
+In client responses we will use URL Encoding to provide additional context so multiple fields worth of
+context can be provided. URL Encoding has [broad language support](https://www.rosettacode.org/wiki/URL_decoding).
+
+### Admin APIs
+
+#### Requests
+
+All Admin APIs that today accept `nil` body should also accept an empty JSON document as request body.
+
+#### Responses
+
+All responses will be JSON objects, a few examples will describe it best. Any error that happens has to be 
+communicated within the originally expected message type. Even the case where JetStream is not enabled for
+an account, the response has to be the data type the API would normally generate with additional `error`.
+
+Successful Stream Info:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "config": {
+    "name": "STREAM",
+    "subjects": [
+      "js.in"
+    ],
+    "retention": "limits",
+    "max_consumers": -1,
+    "max_msgs": -1,
+    "max_bytes": -1,
+    "max_age": 31536000,
+    "max_msg_size": -1,
+    "storage": "file",
+    "num_replicas": 1
+  },
+  "state": {
+    "messages": 95563,
+    "bytes": 40104315,
+    "first_seq": 34,
+    "last_seq": 95596,
+    "consumer_count": 1
+  }
+}
+```
+
+Stream Info Error:
+
+```json
+{
+  "type": "io.nats.jetstream.api.v1.stream_info",
+  "time": "2020-04-23T16:51:18.516363Z",
+  "error": {
```

Comment:
I think the stream info message is:
```javascript
{
   error: {},
   data: {}
}
```

This means that all the fields from the response are correctly packed, not at the top level of the msg. This also implies if there is an error,  that `data` is null.

Reply:
We either need to send the same - valid - document with error set or a completely different document. I am not a fan of the entirely missing/empty data - then it’s a different thing. 

Empty data isn’t valid and can not be validated - so then we have 2 response types from any API. Jsm code can handle that but it’s tedious

Reply:
I am suggesting you do that right inside `data`. And let the response wrapper be generic. I don't know the value of a payload in the error (think go apis), but I am ok with you stuffing whatever - in the end the client can choose to ignore. The point is the envelope containing the response simply has two fields now, rather than surface more fields which probably could have a collision in the future.

Reply:
I prefer not to do that. We use different structs to demarshal, but they all have optional error so I can easily check for that and assume if they are there nothing else is valid. But if its not I can access directly without another level of indirection, and again you are using different structs to demarshal into anyway.

Reply:
Do json schemas not allow for an exclusive or?
Is it not possible to say, when error is present do not have data and status?

Even when this is a jetstream specific envelope that has more than two fields, please don't make me produce empty fields that won't make sense in an error context.
I'm fine with time schema etc... different applications have different fields that make sense for them (server info in monitoring). 

If you change your code from
Config StreamInfo{}  `json:"config"`
to
Config *StreamInfo{}  `json:"config,omitempty"`
assigning StreamInfo{} when all is well and not doing so when there is an error should not be a big one. 
(performance of adding a pointer should be fine compared to json and reflection)

Reply:
For an event like monitoring the things that are *required* would presumably be quite small, so I wouldn't worry too much about your scenario.

For a configuration object defaults and stricter validation is right.

JSON Schema does support all off, any of, one of, but they come at costs - you cant then say for example that undefined fields are denied.

Reply:
The reason for me poking at this is that I believe the established pattern here should be carried out for all our json responses going forward. I fully expect to alter the error to fit the above format and add schema ids/timestamps wherever necessary. 
Below are 3 requests and corresponding responses. (error is still a string, for this example this is besides the point.) 
 
@ripienaar My current reading of this ADR and my understanding of ADR would require me to alter error responses. from response 2 to response 3. Am I understanding ADR and this ADR correctly?

**request1: all is well**
```bash
> nats -s nats://admin:changeit@localhost:4222 req "\$SYS.REQ.SERVER.PING.CONNZ" '{}'
16:02:42 Sending request on [$SYS.REQ.SERVER.PING.CONNZ]
16:02:42 Received on [_INBOX.7AuCQe5rABbXKwx9qBJMms]: '{
  "data": {
    "server_id": "ND26D5AON3567GAFE7ETEGIQOZ4TZPNA3C2VU726JHHK3TRBKCK6ERBR",
    "now": "2020-04-30T16:02:42.610424-04:00",
    "num_connections": 1,
    "total": 1,
    "offset": 0,
    "limit": 1024,
    "connections": [
      {
        "cid": 3,
 ...
        "tls_cipher_suite": "TLS_AES_128_GCM_SHA256"
      }
    ]
  },
  "server": {
    "name": "test",
    "host": "localhost",
    "id": "ND26D5AON3567GAFE7ETEGIQOZ4TZPNA3C2VU726JHHK3TRBKCK6ERBR",
    "cluster": "name",
    "ver": "2.1.6",
    "seq": 15,
    "time": "2020-04-30T16:02:42.610495-04:00"
  }
}'
```

**request 2: with parsing error, current return**
```bash
> nats -s nats://admin:changeit@localhost:4222 req "\$SYS.REQ.SERVER.PING.CONNZ" '{'
16:02:28 Sending request on [$SYS.REQ.SERVER.PING.CONNZ]
16:02:28 Received on [_INBOX.BPL4v7ltvfjZYQazQrmo8j]: '{
  "error": "unexpected end of JSON input",
  "server": {
    "name": "test",
    "host": "localhost",
    "id": "ND26D5AON3567GAFE7ETEGIQOZ4TZPNA3C2VU726JHHK3TRBKCK6ERBR",
    "cluster": "name",
    "ver": "2.1.6",
    "seq": 9,
    "time": "2020-04-30T16:02:28.107164-04:00"
  }
}'
```

**Request 3: parsing error and response according to my understanding of this ADR**
```bash
> nats -s nats://admin:changeit@localhost:4222 req "\$SYS.REQ.SERVER.PING.CONNZ" '{'
16:02:28 Sending request on [$SYS.REQ.SERVER.PING.CONNZ]
16:02:28 Received on [_INBOX.BPL4v7ltvfjZYQazQrmo8j]: '{
  "error": "unexpected end of JSON input",
  "server": {
    "name": "test",
    "host": "localhost",
    "id": "ND26D5AON3567GAFE7ETEGIQOZ4TZPNA3C2VU726JHHK3TRBKCK6ERBR",
    "cluster": "name",
    "ver": "2.1.6",
    "seq": 9,
    "time": "2020-04-30T16:02:28.107164-04:00"
  },
  "data": {
    "server_id": "",
    "now": "",
    "num_connections": 0,
    "total": 0,
    "offset": 0,
    "limit": 0,
    "connections": [],
  }
}'
```

Reply:
I personally do not care for the extra data property. Just make it flat but do same thing as JSON where real response is a ptr with omitempty, So if you error none of that goes on the wire.

Reply:
For an event, advisory or metric number 2 above is fine, I dont care for default things like that in an advisory.

configuring is inherently so requires more care, in JS a unset storage type marshals to `memory` since unset/empty storage type doesnt exist, it defaults.

Reply:
```go
func main() {
	c := api.StreamConfig{}
	cj, _ := json.Marshal(c)

	fmt.Println(string(cj))
}
```

```
{"name":"","retention":"limits","max_consumers":0,"max_msgs":0,"max_bytes":0,"max_age":0,"storage":"memory","discard":"old","num_replicas":0}
```

Now I get that by moving the config into a key and making that a pointer and omit empty you can make this go away, but for configuration or system state I dont see the benefit really. 

I dont think we can/should treat config data and advisory data the same tbh

Reply:
fwiw the way derek implemented it the bodies are empty other than error when there's an error - but the type is still valid and (once I add them) will still have valid and correct data structs.

I wont do any json validation on errors for the API of course - but at least now the door is open for validating responses.